test/test_api_callable.py

```diff
@@ -95,35 +101,126 @@ def test_call_api_callable(self):
+            BackoffSettings(0, 0, 0, 0, 0, 0, 0))
+        mock_grpc.side_effect = CustomException
+        mock_timer.return_value.is_alive.side_effect = [True, True, False]
+        settings = CallSettings(timeout=0, retry=retry)
+        my_callable = api_callable.ApiCallable(mock_grpc, settings)
+        self.assertRaises(CustomException, my_callable, None)
+
+    @mock.patch('grpc.framework.crust.implementations._UnaryUnaryMultiCallable')
+    @mock.patch('google.gax.api_callable.TIMER_FACTORY')
+    def test_retry_times_out_simple(self, mock_timer, mock_grpc):
         to_attempt = 3
         retry = RetryOptions(
-            [Exception],
-            BackoffSettings(None, None, None, 1, None, None, to_attempt))
-        with mock.patch('grpc.framework.crust.implementations.'
-                        '_UnaryUnaryMultiCallable') as mock_grpc:
-            mock_grpc.side_effect = Exception
-            settings = CallSettings(timeout=0, retry=retry)
-            my_callable = api_callable.ApiCallable(mock_grpc, settings)
-            self.assertRaises(Exception, my_callable, None)
-            self.assertEqual(mock_grpc.call_count, to_attempt)
+            [CustomException],
+            BackoffSettings(0, 0, 0, 0, 0, 0, 0))
+        mock_grpc.side_effect = CustomException
+        mock_timer.return_value.is_alive.side_effect = (
+            [True] * to_attempt + [False])
+        settings = CallSettings(timeout=0, retry=retry)
+        my_callable = api_callable.ApiCallable(mock_grpc, settings)
+        self.assertRaises(CustomException, my_callable, None)
+        self.assertEqual(mock_grpc.call_count, to_attempt)
+
+    @mock.patch('grpc.framework.crust.implementations._UnaryUnaryMultiCallable')
+    @mock.patch('google.gax.api_callable.TIMER_FACTORY')
+    def test_retry_aborts_on_unexpected_exception(self, mock_timer, mock_grpc):
+        retry = RetryOptions(
+            [CustomException],
+            BackoffSettings(0, 0, 0, 0, 0, 0, 0))
+        mock_grpc.side_effect = Exception
+        mock_timer.return_value.is_alive.return_value = True
+        settings = CallSettings(timeout=0, retry=retry)
+        my_callable = api_callable.ApiCallable(mock_grpc, settings)
+        self.assertRaises(Exception, my_callable, None)
+        self.assertEqual(mock_grpc.call_count, 1)
+
+    @mock.patch('google.gax.api_callable.TIMER_FACTORY')
+    def test_retry_times_out_no_response(self, mock_timer):
+        retry = RetryOptions(
+            [CustomException],
+            BackoffSettings(0, 0, 0, 0, 0, 0, 0))
+
+        def fake_grpc(*dummy_args, **dummy_kwargs):
+            while True:
+                pass
+        mock_timer.return_value.is_alive.return_value = False
+        settings = CallSettings(timeout=0, retry=retry)
+        my_callable = api_callable.ApiCallable(fake_grpc, settings)
+
+        self.assertRaises(RetryException, my_callable, None)
+
+    @mock.patch('grpc.framework.crust.implementations._UnaryUnaryMultiCallable')
```

Comment:
In most of these tests, the ``call_count`` feature of mock is used; I changed the case where we didn't. Possibly we could use a generic mock, rather than tying it to gRPC. (The mock needs to have arg[0] be the request and arg[1] be the timeout, but it's otherwise arbitrary.)